include/aspect/fe_variable_system.h

```diff
@@ -0,0 +1,423 @@
+
+      /**
+       * Return a vector FeValuesExtractor if this variable is a vector.
+       */
+      const dealii::FEValuesExtractors::Vector &extractor_vector() const;
+
+    private:
+      dealii::FEValuesExtractors::Scalar scalar_extractor;
+      dealii::FEValuesExtractors::Vector vector_extractor;
+  };
+
+
+  /**
+   * A class that represents a collection of variables (of type FEVariable) to
+   * form a Finite Element system.  A dealii::FE_System can be constructed
+   * from it using get_fes() and get_multiplicities().  Other information
+   * contained in this class on top of FE_System are names, and the block
+   * structure for linear systems.
+   */
+  template <int dim>
+  class FEVariableSystem
+  {
+    public:
+      /**
+       * Construct an empty object.
+       */
+      FEVariableSystem();
+
+      /**
+       * Construct object from a vector of variables (identical to calling
+       * initialize()).
+       */
+      FEVariableSystem(const std::vector<VariableDefinition<dim> > &variable_definitions);
+
+      /**
+       * Fill this object with the given list of @p variables.
+       */
+      void initialize(const std::vector<VariableDefinition<dim> > &variable_definitions);
+
+      /**
+       * Return the variable with name @p name. Throws an exception if this
+       * variable does not exist.
+       */
+      const FEVariable<dim> &variable(const std::string &name) const;
+
+      /**
+       * Returns true if the variable with @p name exists in the list of
+       * variables.
+       */
+      bool variable_exists(const std::string &name) const;
+
+      /**
+       * Return the list of all variables.
+       */
+      const std::vector<FEVariable<dim> > &get_variables() const;
+
+      /**
+       * Return the total number of components in the system.
+       */
+      unsigned int n_components() const;
+
+      /**
+       * Return the total number of block of the system.
+       */
+      unsigned int n_blocks() const;
+
+      /**
+       * Return the vector of finite element spaces used for the construction
+       * of the FESystem.
+       */
+      const std::vector<const FiniteElement<dim> *> &get_fes() const;
+
+      /**
+       * Return the vector of multiplicities used for the construction of the
+       * FESystem.
+       */
+      const std::vector<unsigned int> &get_multiplicities() const;
+
+      /**
+       * Return a variable that describes for each vector component which
+       * vector block it corresponds to.
+       */
+      const std::vector<unsigned int> &get_components_to_blocks() const;
+
+    protected:
+      std::vector<FEVariable<dim> > variables;
+
+      std::map<std::string, FEVariable<dim>* > name_map;
+      unsigned int n_components_;
+      unsigned int n_blocks_;
+      std::vector<const FiniteElement<dim> *> fes;
+      std::vector<unsigned int> multiplicities;
+      std::vector<unsigned int> components_to_blocks;
```

Comment:
oh, the pair that grew to a tuple grew by two more members.

not sure the tuple is worth it any more at this time. Instead of structure of arrays, one could envision an array of structures, but I'll leave this up to you.